src/Stratis.Bitcoin.Features.SmartContracts/SmartContractBlockDefinition.cs

```diff
@@ -52,6 +53,13 @@ public sealed class SmartContractBlockDefinition : BlockDefinition
             this.stateRoot = stateRoot;
             this.refundOutputs = new List<TxOut>();
             this.receipts = new List<Receipt>();
+
+            // When building smart contract blocks, we will be generating and adding both transactions to the block and txouts to the coinbase. 
+            // At the moment, these generated objects aren't accounted for in the block size and weight accounting. 
+            // This means that if blocks started getting full, this miner could start generating blocks greater than the max consensus block size.
+            // To avoid this without significantly overhauling the BlockDefinition, for now we just lower the block size by a percentage buffer.
+            // If in the future blocks are being built over the size limit and you need an easy fix, just increase the size of this buffer.
+            this.Options = blockBufferGenerator.GetOptionsWithBuffer(this.Options);
```

Comment:
I'm not sure of the implications of this, but in the `BlockDefinition` constructor it uses the options to set `NeedSizeAccounting`, which is based on the original value of `BlockMaxSize`.

Reply:
For weight-based block size consensus, you may want to track size as well as weight, if the user adds a setting via the command line to have a max block size mined by this miner. Hence `NeedSizeAccounting` is useful to check if a block has ever gone over a certain size in bytes when building it up in weight.

For size-based block size consensus, checking weight and checking size are actually the same calculation - whether `NeedSizeAccounting` is set or not, the result will be identical. We don't actually want to touch that setting in the code I've changed above.

This is because the node is trying to do 2 things at once (Account for weight and size) in the same code. Quite problematic but I will fix this in the FN at some point.